CEXT-1699 Draft version (#131)
AdobeDocs/commerce-extensibility: src/pages/events/convert-field-values.md

---
title: Convert payload field values
description: Learn how to convert a field value based on a condition
keywords:
  - Events
  - Extensibility
---

# Convert payload field values

The payload of an event often includes field values that are not easily interpretable for third-party implementations. For example, a value might be stored in the Commerce database as an integer, but the external system stores the same information as a string. Alternatively, instead of transforming data types, you might want to change a Commerce-supplied string to a string defined in the external system. To address this issue, you can implement a converter, enabling custom values for any field in the payload.

## Converter definition

> Your converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts `mixed` and `Event` arguments following arguments. The method returns a `mixed` data type. Individual converter classes must be created for each field when aiming to convert multiple fields within a payload. Conversely, the same class can be reused to replace a specific field across multiple events.

Author: <!-- Note to developer: Add text that describes advanced usage. Do you have to create a separate class for each field when you want to convert multiple fields within a payload? If you want to replace field X across multiple events, can you reuse the same class? Maybe this is elementary programming, but I'd like to state it. -->

> Your converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts `mixed` and `Event` arguments following arguments. The method returns a `mixed` data type. Individual converter classes must be created for each field when aiming to convert multiple fields within a payload. Conversely, the same class can be reused to replace a specific field across multiple events.

Author: Add text that describes advanced usage. Do you have to create a separate class for each field when you want to convert multiple fields within a payload? If you want to replace field X across multiple events, can you reuse the same class? Maybe this is elementary programming, but I'd like to state it.
 
 `Individual converter classes must be created for each field when aiming to convert multiple fields within a payload. Conversely, the same class can be reused to replace a specific field across multiple events.`

> Your converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts `mixed` and `Event` arguments following arguments. The method returns a `mixed` data type. Individual converter classes must be created for each field when aiming to convert multiple fields within a payload. Conversely, the same class can be reused to replace a specific field across multiple events.

Reviewer: ```suggestion
Your converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts `mixed` and `Event` arguments. The method returns a `mixed` data type. You must create individual converter classes for each field when converting multiple fields within a payload. You can reuse a class to replace a specific field across multiple events.
```

    }
}
```

The default payload for this event includes the following:

```json
{
   "event":{
      "data":{
         "value":{
            "visibility":"4",
        }
      }
   }
}
```

The converter changes the payload to:

```json
{
   "event":{
      "data":{
         "value":{
            "visibility":"CATALOG_AND_SEARCH",
        }
      }
   }
}
```

## Register the converter

You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them using the CLI while subscribing to an event.

### Command line

The [`bin/magento events:subscribe --fields` command](commands.md#subscribe-to-an-event) defines the fields and converters to include in the payload of a subscribed event. The example command adds the `visibility` field and provides the path to the converter class. You can specify multiple fields in the same request.

```bash
bin/magento events:subscribe observer.catalog_product_save_after --fields="store_id" --fields='{"name":"visibility", "converter": "Magento\AdobeCommerceEventsClient\Event\TestConverterVisibility"}'`
```

> ### `config.php` file

Reviewer: Add a blank line above the head3

Add the following line below the head3.

The following example `config.php` is the equivalent of the example `bin/magento events:subscribe` command in the **Command line** example above.

Also make sure there's one blank line above and below the new sentence.

Author: Added

> ```php

Author: Example of config.php file it the event is subscribed using the bash command on [line 106](https://github.com/AdobeDocs/commerce-extensibility/pull/131/files#diff-36693f389686ca3991236ee3437b72b755c9770cb875d50f78f211d1c76b4ec6R106)

'io_events' => [
        'observer.catalog_product_save_after' => [
            'fields' => [
                'store_id',
                [
                    'name' => 'visibility',
                    'converter' => 'Magento\\AdobeCommerceEventsClient\\Event\\TestConverterVisibility'
                ]
            ],
            'enabled' => 1
        ]
 ]       

> ```

Reviewer: Add a blank line above the head3

Author: Added

---

# Convert payload field values

The payload of an event often includes field values that are not easily interpretable for third-party implementations. For example, a value might be stored in the Commerce database as an integer, but the external system stores the same information as a string. Alternatively, instead of transforming data types, you might want to change a Commerce-supplied string to a string defined in the external system. To address this issue, you can implement a converter, enabling custom values for any field in the payload.

## Converter definition

Your converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts `mixed` and `Event` arguments. The method returns a `mixed` data type. You must create individual converter classes for each field when converting multiple fields within a payload. You can reuse a class to replace a specific field across multiple events.

```php
interface FieldConverterInterface
{
    /**
     * Converts a field value
     *
     * @param mixed $value
     * @param Event $event
     * @return mixed
     */
    public function convert(mixed $value, Event $event);
}
```

> As an example, the `observer.catalog_product_save_after` event contains a top-level `visibility` field, which must contain an integer value. You want to convert these values to strings that match values on the external system. The following table describes these values.

Reviewer 2: ```suggestion
As an example, the `observer.catalog_product_save_after` event contains a top-level `visibility` field, which must contain an integer value. Convert these values to strings that match values on the external system. The following table describes these values.
```

```php
<?php
/**
 * Copyright &copy; Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */
declare(strict_types=1);

namespace Magento\AdobeCommerceEventsClient\Event;

use Magento\AdobeCommerceEventsClient\Event\Filter\FieldConverterInterface;

class TestConverterVisibility implements FieldConverterInterface
{
    /**
     * Method used to convert field value
     *
     * @param mixed $value
     * @return mixed
     */
    public function convert(mixed $value, Event $event): mixed
    {
        return match ($value) {
            '1' => 'NOT_VISIBLE_INDIVIDUALLY',
            '2' => 'CATALOG_ONLY',
            '3' => 'SEARCH_ONLY',
            '4' => 'CATALOG_AND_SEARCH'
        };
    }
}
```

The default payload for this event includes the following:

```json
{
   "event":{
      "data":{
         "value":{
            "visibility":"4",
        }
      }
   }
}
```

The converter changes the payload to:

```json
{
   "event":{
      "data":{
         "value":{
            "visibility":"CATALOG_AND_SEARCH",
        }
      }
   }
}
```

## Register the converter

> You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them using the CLI while subscribing to an event.

Reviewer 2: ```suggestion
You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them when using the CLI to subscribe to an event.
```
